### PR TITLE
Randomizer

### PR DESCRIPTION
**Pull request by macfrei**
As per request I build this [team randomizer](https://codesandbox.io/s/team-randomizer-bkn0u) as a react app. It's not perfect yet and this pull request is rather big.

### CODE REVIEW

File: src/GlobalStyles.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { createGlobalStyle } from 'styled-components/macro'
+
+export default createGlobalStyle`
+* {
+    box-sizing: border-box;
+}
+body {
+    display: grid;
+    font-family: sans-serif;
+    grid-template-rows: auto auto 1fr auto;
```
Comment (jamarob) [inline src/GlobalStyles.js:11]: Isn't the first row Reacts #root div and all other components are rendered into it?

File: src/App.js [hunk 2]
```diff
@@ -1,5 +1,35 @@
+import { useState } from 'react'
+import styled from 'styled-components/macro'
+import AppForm from './AppForm'
+import saveToLocal from './services/saveToLocal'
+import loadFromLocal from './services/loadFromLocal'
+import Teams from './Teams'
+
 function App() {
-  return <div>Hello World.</div>
+  const [data, setData] = useState(loadFromLocal('groups') || '')
```
Comment (jamarob) [inline src/App.js]: I was a little bit thrown off by the default empty string and thought data would be of type string

File: src/AppForm.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
+import { useState } from 'react'
+import styled from 'styled-components/macro'
+
+export default function AppForm({ onSubmit }) {
+  const [formData, setFormData] = useState({
+    size: '' || 4,
+    seed: '' || 'Web Development',
+    names: '' || '',
+    prefix: '' || 'HH-WEB-21-1-BR-',
```
Comment (jamarob) [inline src/AppForm.js:10]: `''` is always falsy so every expression evaluates to the right hand side 🤔

File: src/AppForm.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+          />
+        </label>
+        <button>save</button>
+      </Form>
+    </details>
+  )
+
+  function handleSubmit(event) {
+    event.preventDefault()
+    onSubmit(formData)
```
Comment (jamarob) [inline src/AppForm.js]: Maybe add some validation before submitting?

File: src/Teams.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import createGroup from './services/createGroups'
+
+export default function Teams({ data }) {
+  const groups = createGroup(data)
+
+  return (
+    <List>
+      {groups?.map((group, index) => {
+        return (
+          <li key={index}>
```
Comment (jamarob) [inline src/Teams.js:12]: I think using index here might lead to bugs when group size stays the same, not sure.

File: src/Teams.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            <span>
+              {data.prefix}
+              {index + 1}:{' '}
+            </span>
+            {group.sort().map((student, index) => (
+              <>
+                <Name key={index}>{student}</Name>
```
Comment (jamarob) [inline src/Teams.js]: Why not use the name as a key here?

File: src/services/shuffel.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
+export default function shuffle(list, randomizer, seed) {
+  for (let i = list.length - 1; i > 0; i--) {
+    randomizer.seed(seed + i)
```
Comment (jamarob) [inline src/services/shuffel.js:4]: Why do you reseed on every step? Why not pass a correctly preseeded RNG to the function and get rid of the seed argument?

File: src/services/createGroups.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  shuffle(names, rand, seed)
+  let groups = _.chunk(names, Number(size))
+  const lastGroup = groups[groups.length - 1]
+
+  if (lastGroup.length === 1 && groups.length > 1) {
+    groups = [
+      ...groups.slice(0, groups.length - 2),
+      groups[groups.length - 2].slice().concat(lastGroup),
+    ]
+  }
```
Comment (jamarob) [inline src/services/createGroups.js]: What happens when you have groups of 4 and only two students are in the last group?

File: src/GlobalStyles.js (hunk 1 above)
Comment (macfrei) [reply src/GlobalStyles.js:11]: Yes... thank you 👍

File: src/AppForm.js (hunk 4 above)
Comment (macfrei) [reply src/AppForm.js]: Good idea, I'll add some

File: src/services/createGroups.js (hunk 8 above)
Comment (macfrei) [reply src/services/createGroups.js]: Good question! Thanks for pointing that out!

File: src/services/shuffel.js (hunk 7 above)
Comment (macfrei) [reply src/services/shuffel.js:4]: I'll talk to Jerry about this, thank you!

**Pull request merged**